00-programming-fundamentals/intro-to-regex.md

```diff
@@ -0,0 +1,133 @@
+
+## Character Sets
+
+A **character set**, also called a **character class** is a way to tell the regex engine to match only one out of several characters.  We define a character set with square brackets.  For example `/[Ss]/` will match both capital and lowercase S.  Combining the character set with the previous larger literal, `[Aa]da` will match both "Ada" and "ada".  
+
+You can also adjust the character set to accept a range of characters.  For example:  `/[A-Z]/`  will accept a single character in the range A to Z (must be capitalized), while `/[0-9]/` will accept a single digit.  If you wanted to accept any alphabetic characters you could use `/[A-Za-z]/`.
+
+
+**Question:** How could you match any alphanumeric digit?  [Click here to see a solution.](https://gist.github.com/CheezItMan/2d445458f7c25eee37f5fd5b212077f4)
+
+
+## The Wildcard and Special Characters
+
+Sometimes you will want to accept any character.  For that purpose you can use the wildcard character the period (.).  For example the regex `/ada./` will match ada followed by any single character.  Another common need is for characters that can repeat.  For this there are several helpful symbols to use.  These special characters indicate how often a pattern can appear in matching strings.  The plus (+) symbol indicates one or more instances, while the asterix (*) indicates zero or more instances and the question mark (?) indicates a token is optional in that it might appear once or not at all.
+
+| Special Character | Meaning |
+| --- | --- |
+|  .  |  Matches any character |
+|  *  |  Matches if the preceeding token occurs zero or more times.  For example `/[0-9]*/` matches zero or more digits so "",  "123", and "5" would match. |  
+|  +  |  Matches if the preceeding token occurs one or more times.  For example `/[0-9]+/` matches one or more digits so both "1" and "345" would match.  |
+|  ?  |  Marks the preceeding token optional.  So `/example[0-9]?/` matches "example" followed by one or no digits, thus "example", "example1" and "example13" all match.  |
+
+
+Examples:  
+
+- `.+`  - Matches one or more of any characters.  
+-  `/[0-9]+/` - Matches one or more digits
+-  `/[A-Z]*/` - Matches zero or more capital letters.
+-  `/[a-z_][A-Za-z0-9]*/`  - Matches a lower case letter or underscore followed by zero or more alphanumeric characters.  
+
+## Escape characters
+
+There are lots of characters that have special meanings in a Regex (such as the + or * characters).  Just like Strings you can use the backslash character to select the exact character in the text.  It can also be used as a shortcut for common classes of characters.
+
+Some examples include:
+
+| Escape Character |  |  
+| --- | --- |
+| \n  |  newline character |
+| \s  |  Any whitespace character (space, tab, newline) |
+| \S  |  Any non-whitespace character |
+| \d  |  Any digit |
+| \D  |  Any non-digit |
+| \\., \\+, \\*  | The literal character following the backslash  |
```

Comment:
I think we may need just a bit more explanation of what the literal escape character does. Seems like a leap for them to be able to come up with the practice answer without it.